graphql-bal-service/src/main/java/io/ballerina/graphql/schema/generator/SdlSchemaGenerator.java

```diff
@@ -0,0 +1,238 @@
+
+/**
+ * This class implements the GraphQL SDL schema generation.
+ */
+public class SdlSchemaGenerator {
+
+    private static PrintStream outStream = System.out;
+
+    /**
+     * Export the SDL schema for given Ballerina GraphQL services.
+     */
+    public static void generate(Path filePath, Path outPath, String serviceName)
+            throws SchemaGenerationException {
+        Project project = ProjectLoader.loadProject(filePath);
+        PackageCompilation compilation = getPackageCompilation(project);
+        Package packageName = project.currentPackage();
+        DocumentId docId;
+        Document doc;
+        if (project.kind().equals(ProjectKind.BUILD_PROJECT)) {
+            docId = project.documentId(filePath);
+            ModuleId moduleId = docId.moduleId();
+            doc = project.currentPackage().module(moduleId).document(docId);
+        } else {
+            Module currentModule = packageName.getDefaultModule();
+            Iterator<DocumentId> documentIterator = currentModule.documentIds().iterator();
+            docId = documentIterator.next();
+            doc = currentModule.document(docId);
+        }
+
+        SyntaxTree syntaxTree = doc.syntaxTree();
+        SemanticModel semanticModel = compilation.getSemanticModel(docId.moduleId());
+        List<SdlSchema> schemaDefinitions = generateSdlSchema(syntaxTree, semanticModel, serviceName);
+        List<String> fileNames = new ArrayList<>();
+        for (SdlSchema definition : schemaDefinitions) {
+            String fileName = resolveSchemaFileName(outPath, definition.getName());
+            createOutputDirectory(outPath);
+            writeFile(outPath.resolve(fileName), definition.getSchema());
+            fileNames.add(fileName);
+        }
+        if (fileNames.isEmpty()) {
+            return;
+        }
+        outStream.println("SDL Schema(s) generated successfully and copy to :");
+        Iterator<String> iterator = fileNames.iterator();
+        while (iterator.hasNext()) {
+            outStream.println("-- " + iterator.next());
+        }
+    }
+
+    /**
+     * Generate a List of SdlSchema objects for given GraphQL services.
+     */
+    private static List<SdlSchema> generateSdlSchema(SyntaxTree syntaxTree, SemanticModel semanticModel,
+                                                     String serviceName) throws SchemaGenerationException {
+        Map<String, String> servicesToGenerate = new HashMap<>();
+        List<String> availableServices = new ArrayList<>();
+        List<SdlSchema> outputs = new ArrayList<>();
+
+        ModulePartNode modulePartNode = syntaxTree.rootNode();
+        extractSchemaStringsFromServices(serviceName, modulePartNode, semanticModel, availableServices,
+                servicesToGenerate);
+        // If there are no services found for a given service name.
+        if (serviceName != null && servicesToGenerate.isEmpty()) {
```

Comment:
Do we need `serviceName != null` check here?

Reply:
Yes, There can be scenarios where the `serviceName` is not provided and the source file doesn't include any GraphQL services. In that case, we don't need to return the error.

Reply:
Here's my understanding on this:

There can be cases where,
1. Service name is not provided and source file does not include any GraphQL services
2. Service name is not provided and source file include GraphQL services
3. Service name is provided and source file does not include any GraphQL services
4. Service name is provided and source file include GraphQL services

So, from these cases, only 4 should be the valid case, right? Others should be invalid. If that's the case, this will cause a bug. The condition should check for `OR`, instead of `AND`. 

Another thing, if the `serviceName` is not provided, why should the code executes to this far? Shouldn't it return an error early?

Reply:
Here we get the service base path as the `serviceName`. (I changed the variable name to `serviceBasePath` since `serviceName` caused misleading). The service base path is an `optional parameter`. Therefore, we return an error only for case #3. In case #1, it prints a message saying no graphql service found and for case #2, it generates schemas for all the GraphQL services.